File: pixie/eval/rate_limiter.py
"""Central rate limiter for LLM evaluator calls.

Controls throughput using configurable RPS, RPM, TPS, and TPM limits.
Uses asyncio primitives (Semaphore + sliding window) so it integrates
naturally with the async evaluation pipeline.

Usage::

    from pixie.eval.rate_limiter import configure_rate_limits_from_config

    configure_rate_limits_from_config()

The module exposes a singleton via ``get_rate_limiter()``; when no
configuration has been applied the function returns ``None`` and
evaluator calls proceed without throttling.
"""

from __future__ import annotations

import asyncio
import time
from collections import deque

from pixie.config import PixieConfig, RateLimitConfig, get_config

__all__ = [
    "EvalRateLimiter",
    "RateLimitConfig",
    "configure_rate_limits",
    "configure_rate_limits_from_config",
    "get_rate_limiter",
]


class EvalRateLimiter:
    """Central rate limiter for LLM evaluator calls.

    Enforces both request-count and token-count limits using sliding
    windows.  All public methods are async and safe to call from
    concurrent ``asyncio.gather`` tasks.
    """

    def __init__(self, config: RateLimitConfig) -> None:
        self._config = config
        self._lock = asyncio.Lock()

        # Sliding-window records: (timestamp, token_count)
        self._second_window: deque[tuple[float, int]] = deque()
        self._minute_window: deque[tuple[float, int]] = deque()

    @property
    def config(self) -> RateLimitConfig:
        """Return the current configuration."""
        return self._config

    def estimate_tokens(self, text: str) -> int:
        """Estimate token count using ``len(text) // 3`` approximation."""
        return len(text) // 3

    async def acquire(self, estimated_tokens: int = 0) -> None:
        """Wait until the request can proceed within rate limits.

        Blocks (with a short polling interval) until *all* four
        constraints (RPS, RPM, TPS, TPM) are satisfied, then records
        the request in the sliding windows.
        """
        while True:
            async with self._lock:
                now = time.monotonic()
                self._evict(now)

                if self._can_proceed(now, estimated_tokens):
                    self._record(now, estimated_tokens)
                    return

            # Yield control and retry after a short wait
            await asyncio.sleep(0.05)

    # ── internals ────────────────────────────────────────────────────

    def _evict(self, now: float) -> None:
        """Remove entries older than their respective windows."""
        while self._second_window and now - self._second_window[0][0] >= 1.0:
            self._second_window.popleft()
        while self._minute_window and now - self._minute_window[0][0] >= 60.0:
            self._minute_window.popleft()

    def _can_proceed(self, now: float, tokens: int) -> bool:
        """Return ``True`` if all rate-limit constraints are satisfied."""
        sec_count = len(self._second_window)
        sec_tokens = sum(t for _, t in self._second_window)
        min_count = len(self._minute_window)
        min_tokens = sum(t for _, t in self._minute_window)

        if sec_count >= self._config.rps:
            return False
        if min_count >= self._config.rpm:
            return False
        if sec_tokens + tokens > self._config.tps:
            return False
        return min_tokens + tokens <= self._config.tpm

    def _record(self, now: float, tokens: int) -> None:
        """Record a request in both sliding windows."""
        self._second_window.append((now, tokens))
        self._minute_window.append((now, tokens))


# ── module-level singleton ───────────────────────────────────────────────

_rate_limiter: EvalRateLimiter | None = None
_rate_limiter_initialized = False


def configure_rate_limits(config: RateLimitConfig | None = None) -> None:
    """Set (or clear) the global rate limiter for evaluator calls.

    Pass a ``RateLimitConfig`` to enable rate limiting, or ``None`` to
    disable it.

    Args:
        config: The rate-limit configuration.  Passing ``None`` removes
            any previously configured limiter.
    """
    global _rate_limiter, _rate_limiter_initialized  # noqa: PLW0603
    _rate_limiter = None if config is None else EvalRateLimiter(config)
    _rate_limiter_initialized = True


def configure_rate_limits_from_config(config: PixieConfig | None = None) -> None:
    """Apply the central Pixie config to the module-level rate limiter."""
    resolved_config = get_config() if config is None else config
    configure_rate_limits(resolved_config.rate_limits)


def get_rate_limiter() -> EvalRateLimiter | None:
    """Return the active rate limiter, auto-loading it from Pixie config once."""
    if not _rate_limiter_initialized:
        configure_rate_limits_from_config()
    return _rate_limiter

File: pixie-ts/src/eval/rateLimiter.ts
/**
 * Central rate limiter for LLM evaluator calls.
 *
 * Controls throughput using configurable RPS, RPM, TPS, and TPM limits.
 */

import type { PixieConfig, RateLimitConfig } from "../config.js";
import { getConfig } from "../config.js";

/**
 * Sliding-window rate limiter for LLM evaluator calls.
 */
export class EvalRateLimiter {
  private readonly _config: RateLimitConfig;
  private readonly _secondWindow: Array<{ ts: number; tokens: number }> = [];
  private readonly _minuteWindow: Array<{ ts: number; tokens: number }> = [];

  constructor(config: RateLimitConfig) {
    this._config = config;
  }

  get config(): RateLimitConfig {
    return this._config;
  }

  /** Estimate token count using len/3 approximation. */
  estimateTokens(text: string): number {
    return Math.floor(text.length / 3);
  }

  /** Wait until the request can proceed within rate limits. */
  async acquire(estimatedTokens: number = 0): Promise<void> {
    while (true) {
      const now = performance.now() / 1000;
      this.evict(now);

      if (this.canProceed(now, estimatedTokens)) {
        this.record(now, estimatedTokens);
        return;
      }
      await new Promise((resolve) => setTimeout(resolve, 50));
    }
  }

  private evict(now: number): void {
    while (
      this._secondWindow.length > 0 &&
      now - this._secondWindow[0].ts >= 1.0
    ) {
      this._secondWindow.shift();
    }
    while (
      this._minuteWindow.length > 0 &&
      now - this._minuteWindow[0].ts >= 60.0
    ) {
      this._minuteWindow.shift();
    }
  }

  private canProceed(now: number, tokens: number): boolean {
    void now;
    const secCount = this._secondWindow.length;
    const secTokens = this._secondWindow.reduce((sum, e) => sum + e.tokens, 0);
    const minCount = this._minuteWindow.length;
    const minTokens = this._minuteWindow.reduce((sum, e) => sum + e.tokens, 0);

    if (secCount >= this._config.rps) return false;
    if (minCount >= this._config.rpm) return false;
    if (secTokens + tokens > this._config.tps) return false;
    return minTokens + tokens <= this._config.tpm;
  }

  private record(now: number, tokens: number): void {
    this._secondWindow.push({ ts: now, tokens });
    this._minuteWindow.push({ ts: now, tokens });
  }
}

// Module-level singleton
let _rateLimiter: EvalRateLimiter | null = null;
let _rateLimiterInitialized = false;

/**
 * Set (or clear) the global rate limiter for evaluator calls.
 */
export function configureRateLimits(
  config: RateLimitConfig | null = null,
): void {
  _rateLimiter = config ? new EvalRateLimiter(config) : null;
  _rateLimiterInitialized = true;
}

/**
 * Apply the central Pixie config to the module-level rate limiter.
 */
export function configureRateLimitsFromConfig(
  config?: PixieConfig | null,
): void {
  const resolvedConfig = config ?? getConfig();
  configureRateLimits(resolvedConfig.rateLimits);
}

/**
 * Return the active rate limiter, auto-loading from Pixie config once.
 */
export function getRateLimiter(): EvalRateLimiter | null {
  if (!_rateLimiterInitialized) {
    configureRateLimitsFromConfig();
  }
  return _rateLimiter;
}
